File: c/math/0009_palindrom_number/palindrome_number_test.c
#include <stdio.h>
#include <stdbool.h>
#include "palindrome_number.c"

#define RUN_TEST(name, condition)    \
    if (condition)                   \
    {                                \
        printf("[PASS] %s\n", name); \
    }                                \
    else                             \
    {                                \
        printf("[FAIL] %s\n", name); \
    }

int main()
{
    printf("--- Running C Tests ---\n");

    RUN_TEST("Example 1 (Odd length)", isPalindrome(121) == true);
    RUN_TEST("Example 2 (Negative)", isPalindrome(-121) == false);
    RUN_TEST("Example 3 (Ends with 0)", isPalindrome(10) == false);
    RUN_TEST("Single digit zero", isPalindrome(0) == true);

    return 0;
}

File: go/math/0009_palindrome_number/palindrome_number_test.go
package palindromenumber

import (
	"testing"
)

func TestIsPalindrome(t *testing.T) {
	tests := []struct {
		name     string
		value    int
		expected bool
	}{
		{"Example 1 (Odd length)", 121, true},
		{"Example 2 (Negative)", -121, false},
		{"Example 3 (Ends with 0)", 10, false},
		{"Single digit zero", 0, true},
		{"Single digit non-zero", 7, true},
		{"Even length palindrome", 1221, true},
		{"Long odd length palindrome", 1234321, true},
		{"Not a palindrome", 123, false},
		{"Starts and ends with same but not palindrome", 1000021, false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := isPalindrome(tt.value); got != tt.expected {
				t.Errorf("isPalindrome(%d) = %v, want %v", tt.value, got, tt.expected)
			}
		})
	}
}

func BenchmarkIsPalindrome(b *testing.B) {
	benchmarks := []struct {
		name  string
		value int
	}{
		{"SingleDigit", 7},
		{"Negative", -121},
		{"EndsWithZero", 1000},
		{"EvenLength", 1221},
		{"OddLength_Short", 121},
		{"OddLength_Long", 1234321},
		{"NotPalindrome_Long", 1234567},
	}

	for _, bm := range benchmarks {
		b.Run(bm.name, func(b *testing.B) {
			for i := 0; i < b.N; i++ {
				isPalindrome(bm.value)
			}
		})
	}
}
